ecstask/ecs_task.go

```diff
@@ -397,44 +402,121 @@ func removeAFromB(a, b []string) []string {
 	return diff
 }
 
-func deregisterTaskDefinitions(svc *ecs.ECS, taskDefinitionArns []string, parallel int) {
-	arnsChan := make(chan string, len(taskDefinitionArns))
+// Job carries information through a Job channel.
+type Job struct {
+	Arn string
+}
+
+// Result carries information through a Result channel.
+type Result struct {
+	Arn string
+	Err error
+}
+
+func deregisterTaskDefinitions(svc *ecs.ECS, taskDefinitionArns []string, parallel int, debug bool) {
+	worker := func(wg *sync.WaitGroup, jobsChan <-chan Job, resultsChan chan<- Result) {
+		for job := range jobsChan {
+			input := &ecs.DeregisterTaskDefinitionInput{
+				TaskDefinition: aws.String(job.Arn),
+			}
+
+			_, err := svc.DeregisterTaskDefinition(input)
 
-	deregisterTaskDefinition := func(arn string) {
-		_, err := svc.DeregisterTaskDefinition(&ecs.DeregisterTaskDefinitionInput{
-			TaskDefinition: aws.String(arn),
-		})
-		if err != nil {
-			fmt.Println("Error deregistering task definition:", err)
+			result := Result{Arn: job.Arn, Err: err}
+			resultsChan <- result
 		}
+
+		wg.Done()
 	}
 
-	worker := func(wg *sync.WaitGroup) {
-		for arn := range arnsChan {
-			deregisterTaskDefinition(arn)
+	dispatcher := func(wg *sync.WaitGroup, arns []string, parallel int, jobsChan chan Job, resultsChan chan Result) {
+		jobs := stack.New()
+		for _, arn := range arns {
+			jobs.Push(Job{arn})
+		}
+
+		var completedJobs int
+
+		preload := 1
+		if parallel > 1 {
+			preload = parallel - 1
+		}
+
+		for i := 0; i < preload; i++ {
+			jobsChan <- jobs.Pop().(Job)
+		}
+
+		b := &backoff.Backoff{
+			Min:    100 * time.Millisecond,
+			Max:    2 * time.Minute,
+			Jitter: true,
+		}
+
+		for result := range resultsChan {
+			if result.Err != nil {
+				if !isThrottlingError(result.Err) {
+					fmt.Println("\nError deregistering task definition,", result.Err)
+					close(jobsChan)
+					close(resultsChan)
+					wg.Done()
+					os.Exit(1)
+				}
+
+				t := b.Duration()
+				if debug {
+					fmt.Printf("\nbackoff triggered for %s,", result.Arn)
+					fmt.Printf("\nwaiting for %v\n", t)
+				}
+
+				time.Sleep(t)
+				jobs.Push(Job{Arn: result.Arn})
+			} else {
+				b.Reset()
+				completedJobs++
+				fmt.Printf("\r%d deregistered task definitions", completedJobs)
+			}
+
+			if jobs.Len() > 0 {
+				jobsChan <- jobs.Pop().(Job)
+			}
+
+			if completedJobs == len(arns) {
+				close(jobsChan)
+				close(resultsChan)
+			}
 		}
 
 		wg.Done()
 	}
 
-	createWorkerPool := func(numWorkers int) {
-		var wg sync.WaitGroup
-		for i := 0; i < numWorkers; i++ {
-			wg.Add(1)
-			go worker(&wg)
-		}
+	jobsChan := make(chan Job, parallel)
+	resultsChan := make(chan Result, parallel)
 
-		wg.Wait()
+	var wg sync.WaitGroup
+	for i := 0; i < parallel; i++ {
```

Comment:
is allowing `parallel` more of a testing feature or something we actually expect the user to want to configure? It seems like an unnecessary configuration option to me if we can arrive on a sane default.

Reply:
It might be a little more of a testing feature? I dunno, if we allow concurrency I thought it could be nice to provide a lever to turn it off or adjust it.

Reply:
this is definitely a nitpick btw :-D